File: LeetCode/203移除链表元素.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

struct ListNode* removeElements(struct ListNode* head, int val){
    struct ListNode ret, *p, *q;
    ret.next = head;
    p = &ret;
    while (p && p->next) {
        if (p->next->val == val) {
            q = p->next;
            p->next = q->next;
            free(q);
        } else {
            p = p->next;
        }
    }
    return ret.next;
}

// struct ListNode* removeElements(struct ListNode* head, int val){
//     struct ListNode *p = head;
//     if (p == NULL) return head;
//     p->next = removeElements(p->next, val);
//     return (p->val == val)? p->next : p;
// }

File: LeetCode/203移除链表元素.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeElements(ListNode head, int val) {
        ListNode p = new ListNode();
        ListNode q = new ListNode();
        p.next = head;
        q = p;
        while (q != null && q.next != null) {
            if (q.next.val == val) {
                q.next = q.next.next;
            } else {
                q = q.next;
            }
        }
        return p.next;
    }
}